core/modules/date/date.install

```diff
@@ -101,3 +101,18 @@ function date_update_1001() {
   // Remove old system table entries for merged in modules.
   db_query("DELETE FROM {system} WHERE name IN ('date_api', 'date_popup', 'date_views') AND type = 'module'");
 }
+
+/**
+ * Remove entries with null values from date field tables.
+ */
+function date_update_1002() {
+  // Get all date fields.
+  $fields = field_info_fields();
+  foreach ($fields as $field_name => $field) {
+    if (in_array($field['type'], array('date', 'datetime', 'datestamp'))) {
+      db_delete('field_data_' . $field_name)
+        ->isNull($field_name . '_value')
+        ->execute();
+    }
+  }
```

Comment:
We're almost there. One last thing to take care of, is the entity cache (cache_entity_node);

The "NULL" values are still cached. Loading a node will still contain the faulty values, and the nagging's still there.

Reply:
When updates run, the entity caches are cleared. No need to do that again in the update hook.